### DISCUSSION TITLE
Get the list of all nodes with the WebSocket

### DISCUSSION THREAD

**Discussion by Isotech42** [zwave-js/zwave-js-ui#3882] [Q&A] [answered]
Hello !
I maybe have a silly question, but I can't find the command to get the list of all the nodes ids of the controller with the WebSocket.
I've seen that it's a property in the controller that works when I use JS with `controller.nodes`. But I can't find its equivalent for the websocket.

My goal is as follows: I have a dozen Gen5 SmartSwitches and I'm trying to find out their wattage consumption, which I can do on individual nodes with the request : 
```
request = {
        “messageId": message_id,
        “command": ‘node.get_value’,
        “nodeId": 9,
        “valueId": {
            “commandClass": 50, # Meter Command Class
            “property": ‘value’,
            “propertyKey": 66049 # Key of the consumption in W
        }
    } 
```
But I'd like to loop this on all nodes. (writing in Python)

Does anyone have an idea?

**Comment by robertsLando** [comment]
Check docs here: https://github.com/zwave-js/zwave-js-server. I don't maintain that project so I dunno but @raman325 could help you with that

**Reply by Isotech42** [reply]
Yeah I've spent the whole day reading this doc, but I haven't managed to find anything, even when digging through the code. Maybe tiredness and blindness hit again... *grab a cup of coffee*

But yeah, if he's got the solution, I'll take it !

**Reply by robertsLando** [reply]
Tagging also @kpine as he may be able to help you as well

**Reply by kpine** [reply]
I don't quite understand the question. If you want multiple values then just write a Python loop that makes individual queries for each of the values you want. There is not a built-in API to get multiple values if that's what is being asked for.

You get the entire node state in the response when you make the initial connect with `start_listening` command. You can cache that information and dynamically update it when necessary, e.g. by listening to node events, then you'll always have the most recent values. You can grab the entire node state again with `endpoint.get_node_unsafe`, but doing this often might be expensive. Otherwise, if you want to manually refresh, do the looping yourself.

Since you're using Python, I'd suggest to take advantage of the existing pylib and use it instead of writing the websocket API code yourself. If you don't want that dependency, then you can use its code as inspiration and to answer your questions. https://github.com/home-assistant-libs/zwave-js-server-python

If the question is a more general "how to write the Python code", well that's a broad question to answer. See the pylib for examples and tests, and then HA's [Z-Wave integration](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zwave_js) for more pylib usage examples.

**Reply by Isotech42** [reply]
My problem isn't the fact of making several requests to the nodes or the fact of writing it in Python, but rather having the list of nodes registered on the controller.

In fact, the code I've written above sends a request to just 1 node with id X to find out the consumption status of this node X.
But as I don't know the ids of the nodes connected to the controller (given that the ids are likely to change), I'm trying to make an initial request to find out the list of ids to store in order to loop through them.

The pseudo-code I want to achieve is more as follows:

```python
# First request to send
request = {
    "messageId" : "all_nodes_ids"
    "command" : "controller.get_all_nodes_ids"    # <--- The command I'm searching
}
websocket.send(request)

# Receive the response
def on_message(websocket, message):
    data = json.loads(message)
    if data.get("messageId") == "all_nodes_ids" :
        global all_nodes_ids = data.get("nodes")    # Likely to be a list of all the nodes registered like [1,3,4,5,7,8,10,11,12,13]
    
# Then sending to all nodes the request to get the consumption
for node_id in all_nodes_ids:
        message_id = str(uuid.uuid4())
        request = {
            "messageId": message_id,
            "command": "node.get_value",
            "nodeId": node_id,
            "valueId": {
                "commandClass": 50,
                "property": "value",
                "propertyKey": 66049 
            }
        }
        ws.send(request)
```

It's really the command to get the list of all nodes registered on the controller that I'm looking for (used for the loop after).

However, if this list is already sent when you do a `start_listening` and I haven't seen it, that's possible too. Or if there's another way to make the request in broadcast to all nodes or something like that, I'd be interested too.

(By the way, I hadn't seen that there was a Python wrapper for websocket, I'll go and check it out!)

Thank's !

**Reply by kpine** [reply, marked as answer]
You're correct, there's no specific command to get a list of node IDs, that doesn't even exist in the Z-Wave JS API. In JS you would just access the controller's [`nodes` property](https://zwave-js.github.io/node-zwave-js/#/api/controller?id=nodes) which returns the nodes as objects. The most equivalent for the WS API would be requesting the [controller state](https://github.com/zwave-js/zwave-js-server/?tab=readme-ov-file#get-controller-state), and the response embeds the entire node states. A node state is all the node's properties along with all its values. This can be quite a bit of data for large networks.

In general, the server APIs are roughly one-to-one with the Z-Wave JS API and commands are mostly the methods, not properties.

The expectation is that you use `start_listening` in the [connect sequence](https://github.com/zwave-js/zwave-js-server/?tab=readme-ov-file#api) to get the bulk list of nodes and values and cache that, then listen for notification events to get updates in state, e.g. nodes when nodes are added or removed, values added or updated, etc.

For your use case, if this is just a one-off command (e.g. a CLI) you could skip the `node.get_value` command and just access the values from the node dump obtained from the controller state. Of course, since this is an async system, the values are possibly outdated by the time you get the data back and process it. But if you have anything that's running long-term, you'd want to use events instead as getting the full state every time will be expensive.

The python library handles all of the events and caching the state for you. The only thing is the library is async so it's a little more difficult to work with.

**Reply by robertsLando** [reply]
an alternative could be to send a driver function to return required info

**Reply by Isotech42** [reply]
Thank you for your answers!
My use-case is actually finding a way of using just a CLI command to get the consumption of all SmartSwitch at a given time, without necessarily needing a connected client for the events.

I've tried retrieving the `start_listening` response and I get all the info I need (which I just need to clean up to keep what I'm interested in). So I can skip the loop with the `node.get_value` as explained by @kpine .
> For your use case, if this is just a one-off command (e.g. a CLI) you could skip the node.get_value command and just access the values from the node dump obtained from the controller state.


By the way, the Python library works really well for retrieving events, but if the SmartSwitches haven't changed their value, I can't see them in the terminal (obviously). I can use `--dump-state` which  is the same as sending the `start_listening` command and closing the connection.

**Reply by Isotech42** [reply]
If anyone is interested in how I did it :

```python
def on_open(ws):
    print("Connection opened. Sending request...")

    message_id = str(uuid.uuid4())
    request = {
        "messageId": message_id,
        "command": "start_listening",
    }
    ws.send(json.dumps(request))

def on_message(ws, message):
    data = json.loads(message)

    # Check if we received the controller state
    if 'result' in data and 'state' in data['result']:
        nodes = data['result']['state']['nodes']
        for node in nodes:
            node_id = node.get('nodeId')
            values = node.get('values', [])
            for value in values:
                # Check if it's the Meter Command Class (50) and propertyKey for watts (66049)
                if value.get('commandClass') == 50 and value.get('propertyKey') == 66049:
                    power_consumption = value.get('value')
                    print(f"Node {node_id} Power Consumption: {power_consumption} W")
        # Close connection after retrieving the data
        ws.close()
```

Result :

```
Starting WebSocket connection...
Connection opened. Sending request...
Node 7 Power Consumption: 2.578 W
Node 8 Power Consumption: 1.845 W
Node 9 Power Consumption: 1.947 W
Node 10 Power Consumption: 3.309 W
Node 11 Power Consumption: 0 W
Node 12 Power Consumption: 0 W
Node 13 Power Consumption: 1.687 W
Node 14 Power Consumption: 2.457 W
Node 15 Power Consumption: 1.965 W
Node 16 Power Consumption: 0.579 W
Node 17 Power Consumption: 0 W
Node 18 Power Consumption: 0 W
Node 19 Power Consumption: 3.425 W
WebSocket connection closed.
```
